lib/timers/promises.js

```diff
@@ -111,7 +115,59 @@ function setImmediate(value, options = {}) {
       () => signal.removeEventListener('abort', oncancel)) : ret;
 }
 
+async function* setInterval(after, value, options = {}) {
+  validateObject(options, 'options');
+  const { signal, ref = true } = options;
+  validateAbortSignal(signal, 'options.signal');
+  validateBoolean(ref, 'options.ref');
+
+  if (signal?.aborted)
+    throw new AbortError();
+
+  let onCancel;
+  let interval;
+  try {
+    let notYielded = 0;
+    let callback;
+    interval = new Timeout(() => {
+      notYielded++;
+      if (callback) {
+        callback();
+        callback = undefined;
+      }
+    }, after, undefined, true, true);
+    if (!ref) interval.unref();
+    insert(interval, interval._idleTimeout);
+    if (signal) {
+      onCancel = () => {
+        // eslint-disable-next-line no-undef
+        clearInterval(interval);
+        if (callback) {
+          callback(PromiseReject(new AbortError()));
+          callback = undefined;
+        }
+      };
+      signal.addEventListener('abort', onCancel, { once: true });
+    }
+
+    while (!signal?.aborted) {
+      if (notYielded === 0) {
+        await new Promise((resolve) => callback = resolve);
+      }
+      for (; notYielded > 0; notYielded--) {
+        yield value;
```

Comment:
If `typeof value === 'function'`, then it would be worthwhile calling the function here. Also, it would be worthwhile making this an `await` so that if `value` is a promise the yield happens after the Promise is resolved...

So, something like...
```suggestion
        yield await (typeof value === 'function' ? value() : value);
```

This would allow patterns such as:

```js
async function getValue() { /** ... **/ }

for await (const a of setInterval(100, getValue)) {
  console.log(a);
}

// and

for await (const a of setInterval(100, getValue())) {
  console.log(a);
}
```

Reply:
@jasnell 
Ideally when [this](https://github.com/tc39/proposal-iterator-helpers) lands (stage 2, so good chance) you could just do `setInterval(100).map(getValue)` which is kind of neater :]